### PR TITLE
Add math methods

### LINKED ISSUE #687: [Feature Request] _.sum
I know that you can use _.reduce to calculate the sum of an collection, but I think that the sum is a very basic and often needed operation. Thus, a shorthand for it would be appropriate.

Compare the length of the following to lines:

``` javascript
_.reduce(array, function(sum, num) { return sum + num}, 0);
_.sum(array);
```

It could provide a _.pluck style shorthand for summing specific attributes:

``` javascript
_.reduce(array, function(sum, num) { return sum + num.attribute}, 0);
_.sum(array, "attribute");
```

I for myself often find myself trying to use _.sum and then I'm always surprised that it doesn't exist. Do you agree? Or is it inappropriate for lodash?

### PR DESCRIPTION
**Pull request by thejameskyle**
Resolves #687 and roadmap item

### CODE REVIEW

File: lodash.src.js [hunk 1]
```diff
@@ -11156,6 +11156,211 @@
...
+     * @param {number} exponent
+     * @returns {number}
+     * @example
+     *
+     * _.power(6, 4);
+     * // => 1296
+     */
+    function power(base, exponent) {
+      return Math.pow(base, exponent);
+    }
```
Comment (megawac) [inline lodash.src.js]: what's the point of this one

File: lodash.src.js [hunk 2]
```diff
@@ -11156,6 +11156,211 @@
...
+     * // => 4
+     *
+     * _.round(4.006, 2);
+     * // => 4.01
+     *
+     * _.round(4060, -2);
+     * // => 4100
+     */
+    function round(number, precision) {
+      if (precision) {
```
Comment (megawac) [inline lodash.src.js]: just default to 0? save a couple loc

File: lodash.src.js [hunk 3]
```diff
@@ -11156,6 +11156,211 @@
...
+     * @example
+     *
+     * _.median([4, 6, 2]);
+     * // => 4
+     *
+     * _.median([4, 6, 2, 2]);
+     * // => 3
+     */
+    function median(array) {
+      if (!isArray(array)) return;
```
Comment (megawac) [inline lodash.src.js]: on all of these I'd say returning NaN would be most appropriate

**Comment by thejameskyle** [comment]
Damn it travis

File: lodash.src.js (hunk 2 above)
Comment (thejameskyle) [inline lodash.src.js]: This was to avoid the extra calculations for 0 and other falsy values.

File: lodash.src.js (hunk 1 above)
Comment (thejameskyle) [inline lodash.src.js]: Completeness I guess, lodash-math had it.

File: lodash.src.js (hunk 3 above)
Comment (thejameskyle) [inline lodash.src.js]: :+1:

Comment (thejameskyle) [inline lodash.src.js]: Would that include `_.median(1);`? Also, is it worth supporting `_.median(1, 2, 3);`?

Comment (megawac) [inline lodash.src.js]: same for empty arrays

**Comment by megawac** [comment]
For the record I think these belong in separate packages such as mathp and _.math

File: lodash.src.js [hunk 4]
```diff
@@ -11156,6 +11156,212 @@
...
+     * @param {number} array The numbers to add.
+     * @returns {number} Returns the sum.
+     * @example
+     *
+     * _.sum([4, 6, 2]);
+     * // => 12
+     */
+    function sum(array) {
+      if (!isArray(array) || array.length === 0) return NaN;
+      return reduce(array, add);
```
Comment (megawac) [inline lodash.src.js]: you'll want to put the initval to 0

**Comment by phated** [comment]
Yay math

**Comment by jdalton** [comment]
The travis fails are being trackered [here](https://github.com/travis-ci/travis-ci/issues/2838#issuecomment-75417443) (kind of) not sure what's causing them.

File: lodash.src.js [hunk 5]
```diff
@@ -11156,6 +11156,212 @@
...
+     * _.median([4, 6, 2]);
+     * // => 4
+     *
+     * _.median([4, 6, 2, 2]);
+     * // => 3
+     */
+    function median(array) {
+      if (!isArray(array) || array.length === 0) return NaN;
+      var middle = (array.length + 1) / 2,
+          sorted = baseCreate(array).sort();
```
Comment (thejameskyle) [inline lodash.src.js]: reminder to self: test this `baseCreate`

**Comment by d10** [comment]
I don't think add, subtract, multiply, divide and modulo are required, as they are supported natively, and adding them would just cause decreased performance as compared to the native operators.

**Comment by jdalton** [comment]
Let's go with the less is more approach here since Math methods can snowball and a line to draw is kind of blurry. Let's start with the most requested one `sum` and `add` (as it's the base). Then we can add others as they are requested.

**Comment by jdalton** [comment]
> I don't think add, subtract, multiply, divide and modulo are required, as they are supported natively, and adding them would just cause decreased performance as compared to the native operators

It's about function composition here, since there is no `add` method and the like that are built-in.

**Comment by thejameskyle** [comment]
@d10 You wouldn't use something like `_.add(1, 2)` in place of `1 + 2`, you'd use it together with other methods in a chain or in partials. For example a more functional style version of the sum method included above:

```js
var sum = _.partial(_.reduce, _, _.add);
```

@jdalton Can I keep `_.round` too _pleeeease_? The precision is super useful. We do stuff like that at CloudFlare constantly converting bytes/requests/etc.

**Comment by jdalton** [comment]
>  Can I keep _.round too pleeeease? The precision is super useful. We do stuff like that at CloudFlare constantly converting bytes/requests/etc.

I think we should wait for the demand to be there. I use `ceil` or `floor` more ;)

**Comment by thejameskyle** [comment]
DEMAND

jk, will remove shortly

**Comment by jdalton** [comment]
> DEMAND

On second look demand looks a little tiny :P

**Comment by jdalton** [comment]
Your PR appears to be against 3.2.0 too so that'll need to be fixed too.

**Comment by jdalton** [comment]
PR needs rebase.

File: lodash.src.js (hunk 4 above)
Comment (Droogans) [inline lodash.src.js]: I thought the idea here wasn't to [call other lodash functions](https://github.com/lodash/lodash/issues/687#issuecomment-74453969) directly?

Comment (jdalton) [inline lodash.src.js]: > I thought the idea here wasn't to call other lodash functions directly?

It won't be, but I rather not nit the PR to death and get it in, then tweak after the fact.

Comment (thejameskyle) [inline lodash.src.js]: I'll fix this right now

Comment (thejameskyle) [inline lodash.src.js]: Actually is it even worth having `_.add` with this change? All it does it make the code less concise.

**Comment by thejameskyle** [comment]
Alright, everything should be good. I kept a copy of the other methods if you ever want them in the future.

**Comment by psirenny** [comment]
Hey, I was just wondering why modulo didn't get any love here. Should I create an issue for it and see if it gets some 👍 ?

**Comment by jdalton** [comment]
@psirenny totally send in a PR.

**Pull request merged**